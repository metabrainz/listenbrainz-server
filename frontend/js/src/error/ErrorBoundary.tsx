import * as React from "react";
import { Helmet } from "react-helmet";
import { isRouteErrorResponse, Link, useRouteError } from "react-router-dom";

const ErrorStatusMessage: { [key: number]: string } = {
  400: "Invalid request",
  401: "Unauthorized",
  403: "Access denied",
  404: "Page not found",
  413: "Filesize limit exceeded",
  500: "Internal server error",
  503: "Service unavailable",
};

type RouteError = {
  status?: number;
  message?: string;
  data?: any;
};

export function ErrorBoundary() {
  const error = useRouteError() as RouteError;
  const errorStatus = error.status || 500;
  const errorStatusMessage = ErrorStatusMessage[errorStatus] || "Error";

  const errorMessage = error.data?.error || error.message || errorStatusMessage;

  return isRouteErrorResponse(error) ? (
    <>
      <Helmet>
        <title>{errorStatusMessage}</title>
      </Helmet>
      <h2 className="page-title">{errorStatusMessage}</h2>
      <p>
        <code>{`${errorStatus}: ${errorMessage}`}</code>
      </p>
      <p>
        <Link to="/">Back to home page</Link>
      </p>
    </>
  ) : (
    <>
      <Helmet>
        <title>Error</title>
      </Helmet>
      <h2 className="page-title">An error occured</h2>
      <p>{errorMessage}</p>
    </>
  );
}

export default ErrorBoundary;
