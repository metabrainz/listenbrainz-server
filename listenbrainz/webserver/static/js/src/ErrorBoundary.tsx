import * as React from "react";

export type ErrorBoundaryState = {
  hasError: boolean;
  error: Error | null;
};

export default class ErrorBoundary extends React.Component<
  {},
  ErrorBoundaryState
> {
  constructor(props: {}) {
    super(props);

    this.state = { hasError: false, error: null };
  }

  componentDidCatch(error: Error) {
    // Update state so the next render will show the fallback UI.
    this.setState({ hasError: true, error });
  }

  render() {
    const { children } = this.props;
    const { hasError, error } = this.state;

    if (hasError) {
      document.title = "Something went wrong - ListenBrainz";
      return (
        <div>
          <h2 className="page-title">{error!.name}</h2>
          <p>{error!.message}</p>
          <p>
            <button
              className="btn-link"
              type="button"
              onClick={() => {
                window.location.reload();
              }}
            >
              Reload the page
            </button>
          </p>
        </div>
      );
    }

    return children;
  }
}
