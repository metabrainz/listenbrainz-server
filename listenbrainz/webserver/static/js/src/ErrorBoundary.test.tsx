import * as React from "react";
import { mount } from "enzyme";

import ErrorBoundary from "./ErrorBoundary";

const ChildComponent = () => {
  return <div>Child Component</div>;
};

describe("ErrorBoundary", () => {
  it("renders error page correctly if child component throws error", () => {
    const wrapper = mount(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>
    );

    wrapper.find(ChildComponent).simulateError(Error("Test Error"));

    expect(wrapper.state()).toMatchObject({ hasError: true });
    expect(wrapper).toMatchSnapshot();
  });

  it("reloads page when button is clicked", () => {
    const wrapper = mount(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>
    );

    wrapper.find(ChildComponent).simulateError(Error("Test Error"));
    // Mock the reload function
    Object.defineProperty(window, "location", {
      value: {
        reload: jest.fn(),
      },
      writable: true,
    });
    wrapper.find("button").simulate("click");

    expect(window.location.reload).toHaveBeenCalledTimes(1);
  });

  it("renders the child component if error is not thrown", () => {
    const wrapper = mount(
      <ErrorBoundary>
        <ChildComponent />
      </ErrorBoundary>
    );

    expect(wrapper.state()).toMatchObject({ hasError: false });
    expect(wrapper).toMatchSnapshot();
  });
});
