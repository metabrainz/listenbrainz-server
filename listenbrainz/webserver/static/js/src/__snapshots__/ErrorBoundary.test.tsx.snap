// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders error page correctly if child component throws error 1`] = `
<ErrorBoundary>
  <div>
    <h2
      className="page-title"
    >
      Error
    </h2>
    <p>
      Test Error
    </p>
    <p>
      <button
        className="btn-link"
        onClick={[Function]}
        type="button"
      >
        Reload the page
      </button>
    </p>
  </div>
</ErrorBoundary>
`;

exports[`ErrorBoundary renders the child component if error is not thrown 1`] = `
<ErrorBoundary>
  <ChildComponent>
    <div>
      Child Component
    </div>
  </ChildComponent>
</ErrorBoundary>
`;
